diff.py

```diff
@@ -475,10 +476,71 @@ def __init__(self, value: int, ansi_code: str):
 FormatFunction = Callable[[str], Format]
 
 
+@dataclass
+class Text:
+    segments: List[Tuple[str, Format]]
+
+    @staticmethod
+    def formatted(line: str, f: Format) -> "Text":
+        return Text([(line, f)])
+
+    @staticmethod
+    def raw(line: str) -> "Text":
+        return Text.formatted(line, Format.NONE)
+
+    @staticmethod
+    def empty() -> "Text":
+        return Text([])
+
+    def reformat(self, f: Format) -> "Text":
+        return Text.formatted(self.plain(), f)
+
+    def plain(self) -> str:
+        return "".join(s for s, f in self.segments)
+
+    def __str__(self) -> str:
+        # Use Formatter.apply(...) instead
+        return NotImplemented
+
+    def __repr__(self) -> str:
+        return f"<Text: {self.plain()!r}>"
+
+    def __add__(self, other: Union["Text", str]) -> "Text":
+        if isinstance(other, str):
+            return self + Text.raw(other)
+        return Text(self.segments + other.segments)
+
+    def __radd__(self, other: Union["Text", str]) -> "Text":
+        if isinstance(other, str):
+            return Text.raw(other) + self
+        return Text(other.segments + self.segments)
+
+    def finditer(self, pat: Pattern[str]) -> Iterator[Match[str]]:
+        """Replacement for `pat.finditer(text)` that operates on the inner text,
+        and returns the exact same matches as `Text.sub(pat, ...)`."""
+        for chunk, f in self.segments:
+            for match in pat.finditer(chunk):
+                yield match
+
+    def sub(self, pat: Pattern[str], sub_fn: Callable[[Match[str]], "Text"]) -> "Text":
+        new_segments: List[Tuple[str, Format]] = []
+        for chunk, f in self.segments:
+            i = 0
+            for match in pat.finditer(chunk):
+                start, end = match.start(), match.end()
+                assert i <= start <= end <= len(chunk)
+                sub = sub_fn(match)
+                new_segments.append((chunk[i:start], f))
+                new_segments.extend(sub.segments)
+                i = end
+            new_segments.append((chunk[i:], f))
+        return Text(new_segments)
```

Comment:
Heh, it's funny how this illuminates how much sketchy stuff we were doing on formatted text. It would be nice to refactor the code to only do this kind of processing on `str`'s, and also get rid of a bunch of `.plain()` and `.reformat()` in the process. But I totally understand if you don't want to dig into that mess.

Reply:
Yeah, I think a nicer approach would involve tokenizing the lines first? And then using the regexes to match tokens, rather than using `re.sub(...)`-style operations. 
Tokenizing might be easy -- it be a simple as splitting on `\t` and `,`? But overall that seemed like a bigger change and I was nervous about getting it right for all 3 supported arches.

Reply:
Tokenizing would be nice for sure, but yeah, it's a larger change and it may be a good idea to build up a small test suite first... What I was hoping for was more of a smaller change that keeps `out1`, `out2` as `str` for long enough that regexes-on-Text wouldn't be necessary